### ISSUE TITLE
Proposal: Dynamic and late-mounting custom elements

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/can-stache#360]
__tldr;__ `can-stache` and `can-component` should support dynamic custom elements.  For example, this would allow something like:

```
<{{tagName}}>
```

To change which component is rendered when the `tagName` variable changes.

## Proposal 

I propose supporting magic tags within a tag like `<{{tagName}}>`.  

 - If `tagName` is evaluated to a `String` like `"my-component"`, `my-component` will be looked up in `can-view-callbacks`.  If a callback is not found, a warning will be given.
 - If `tagName` is evaluated to be a Constructor function, that `Constructor` function will be called with `new Constructor(el,` [tagData](https://canjs.com/doc/can-view-callbacks.tagData.html) `)`.
- If `tagName` is evaluated to `null` or `undefined`, a warning will be given.
- If `tagName` is a `renderer` function, that renderer function will be rendered with the data created by the `can-stache-bindings`.

Additionally, as the value of `tagName` changes, the page will be updated accordingly.

Finally, by supporting all of the previous steps, we should easily be able to support late-mounting of custom elements called with the existing syntax.  For example, if someone were to define a component after the rendering of a template that uses it:

```js
var template = stache(`<foo-bar/>`);
document.body.appendChild(`<foo-bar/>`)
Component.extend({
  view: stache('hello world')
})
```

We should be able to mount this custom element.

> NOTE: This form of mounting differs from the MutationObserver form in 4.0.  This mounting will use a event listener on `can-view-callbacks`.

## Problems and Solutions

There are 3 problems this issue seeks to address:

1. Dynamic custom elements
2. Reducing the importance of `can-view-callbacks`
3. Order of definition problems

### Dynamic Custom Elements

Currently, applications like Bitballs [have to do weird hacks to support rendering dynamic elements](https://github.com/donejs/bitballs/blob/master/public/app.js#L197):

```js
stache.registerHelper("pageComponent", function(options){
	var pageComponent = options.context.pageComponentConfig,
		template =
			"<can-import from='bitballs/components/" + pageComponent.moduleName + "'>" +
				"{{#if isResolved}}" +
					"{{#with ../.}}<"+pageComponent.componentName + " " + pageComponent.attributes + "/>{{/with}}" +
				"{{else}}" +
					"Loading..." +
				"{{/if}}" +
			"</can-import>";
	return stache(template)(this, options.helpers, options.nodeList);
});
```

This is a common pattern.  Simplifying it would also improve the [Routing Guide](https://github.com/canjs/can-route/issues/122).

### Reducing the importance of `can-view-callbacks`

`can-view-callbacks` is a "stateful" module.  While this proposal doesn't seek to outright eliminate it, we could begin to go that direction if it becomes possible to pass components as follows:

```html
<can-import from="my-component" value:to="scope.vars.MyComponent"/>
<{{scope.vars.MyComponent}}>
```

In this case, `my-component.js` would not actually have to use `tag`, instead it could look like:

```js
export default Component.extend({
  view: stache("Hi There")
})
```

## Benefits

**Comment by frank-dspeed** [comment]
Is that not implemented via.can import inside a switch statement already?

**Comment by justinbmeyer** [comment]
That’s progressive loading, not late binding.

**Comment by frank-dspeed** [comment]
@justinbmeyer could you maybe help me i have at present exactly a case where i need to understand can-view-callbacks :)

value:to="scope.vars.MyComponent" would help me probally a lot i try to render a stach template with with can import in it but some how it assigns that import not to the same can-view-callback instance i think

because i get always no can-view-callback element registered. but also no fail of can import all is loading

**Comment by justinbmeyer** [comment]
Can you ask this question elsewhere (forums or a can-view-callbacks issue) and be more specific about the problem (I’m not sure what you mean).

**Comment by rjgotten** [comment]
The major problem with the 'weird hack' seems to be the fact that attributes and view model bindings may differ depending on which component is dynamically loaded into place.

How'd you support varying attirbutes and bindings when only the tag-name is observed in a construct like `<{{tag-name}}>` ?

Wouldn't you be better off with a tag callback, like `<can-import>` to handle this? E.g.

```<can-component component:from="pageComponent" />```

where `pageComponent` is some kind of spec object holding the component's constructor; light DOM template; as well as attributes/bindings. E.g.

```js
{
  Component : MyComponent,
  template  : stache( "<span>Light DOM content</span>" )
  bindings  : {
    "from:foo" : "bar",
    "from:bar" : "baz",
    // ...
  }
}
```

`Component` could also be a promise here, or a string serving as a module ID to download -- which would effectively turn it into a promise and then send it down the same code path.

The light DOM content templates for resolved components are handled as part of the spec object, which makes them capable of being injected dynamically as well. And it frees up the inner content of the `<can-component>` tag to be used as the placeholder template for the loading state:

```mustache
<can-component component:from="pageComponent">
  <span>Loading...</span>
</can-component>
```


(And all this you could probably get working with just the elementary building blocks that have been available as far back as CanJS 2.x ...)

**Comment by justinbmeyer** [comment]
> The major problem with the 'weird hack' that requires most of its heavy-handedness seems to be the fact that attributes, view model bindings and light DOM templates may differ depending on which component is dynamically loaded into place.

Agreed, but I think that there's still a common use case where every component has the same template and bindings:

```
<{{main-component}} appVM:from="this"/>
```

or if we eventually support assign / update:

```
<!-- similar to JSX's {...props} -->
<{{main-component}} appVM:assignFrom="this"/>
```

If there's a big switch case, with different bindings and light DOM, I sorta see it as a related-but-separate concern.  I think `<{{tag}}>` is long overdue.

**Comment by justinbmeyer** [comment]
@rjgotten part of the problem I think is that a big SWITCH feels SOOOOO heavy in `can-stache`.
